File: springboot-boilerplate/src/main/java/com/grepp/spring/app/controller/web/member/MemberController.java
package com.grepp.spring.app.controller.web.member;

import com.grepp.spring.app.controller.web.member.payload.SigninRequest;
import com.grepp.spring.app.controller.web.member.payload.SignupRequest;
import com.grepp.spring.app.model.auth.code.Role;
import com.grepp.spring.app.model.member.MemberService;
import com.grepp.spring.app.model.member.dto.MemberDto;
import com.grepp.spring.infra.error.exceptions.CommonException;
import com.grepp.spring.infra.response.ResponseCode;
import jakarta.servlet.http.HttpSession;
import jakarta.validation.Valid;
import java.util.UUID;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.Authentication;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.validation.BindingResult;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.servlet.mvc.support.RedirectAttributes;

@Controller
@RequiredArgsConstructor
@Slf4j
@RequestMapping("member")
public class MemberController {

    private final MemberService memberService;

    @GetMapping("signup")
    public String signup(SignupRequest form){
        return "member/signup";
    }

    @PostMapping("verify")
    public String verifySignup(
        @Valid SignupRequest form,
        BindingResult bindingResult,
        HttpSession session,
        RedirectAttributes redirectAttributes){

        if(bindingResult.hasErrors()){
            return "member/signup";
        }

        String token = UUID.randomUUID().toString();
        MemberDto dto = form.toDto();
        memberService.sendVerificationMail(token, dto);
        session.setAttribute(token, dto);
        redirectAttributes.addAttribute("msg", "회원가입 메일이 발송되었습니다.");
        return "redirect:/";
    }

    @GetMapping("signup/{token}")
    public String signup(
        @PathVariable
        String token,
        HttpSession session
    ){

        MemberDto dto = (MemberDto) session.getAttribute(token);

        if(dto == null){
            throw new CommonException(ResponseCode.INVALID_TOKEN);
        }

        memberService.signup(dto, Role.ROLE_USER);
        session.removeAttribute(token);
        return "redirect:/";
    }

    @GetMapping("signin")
    public String signin(SigninRequest form){
        return "member/signin";
    }


    @GetMapping("mypage")
    public String mypage(Authentication authentication, Model model){
        log.info("authentication : {}", authentication);
        String userId = authentication.getName();
        MemberDto memberDto = memberService.findById(userId);
        model.addAttribute("member", memberDto);
        return "member/mypage";
    }

    @PreAuthorize("hasAuthority('ROLE_ADMIN') or authentication.name == #id")
    @GetMapping("{id}")
    public String get(@PathVariable String id, Model model){
        MemberDto memberDto = memberService.findById(id);
        model.addAttribute("member", memberDto);
        return "member/mypage";
    }
}

File: user-service/src/main/kotlin/com/grepp/spring/app/controller/api/member/MemberController.kt
package com.grepp.spring.app.controller.api.member

import com.grepp.spring.app.controller.api.member.payload.MemberDetailResponse
import com.grepp.spring.app.controller.api.member.payload.SignupRequest
import com.grepp.spring.app.model.member.MemberService
import com.grepp.spring.infra.response.ApiResponse
import jakarta.servlet.http.HttpSession
import jakarta.validation.Valid
import org.springframework.http.ResponseEntity
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("api/member")
class MemberController(
    private val memberService: MemberService
) {

    @GetMapping("add/{token}")
    fun signup(
        @PathVariable
        token:String,
        session:HttpSession
    ):ResponseEntity<ApiResponse<Unit>>{
        val request = session.getAttribute(token) as SignupRequest
        memberService.signup(request)
        return ResponseEntity.ok(ApiResponse.noContent())
    }

    @PostMapping("verify")
    fun verify(
        @Valid
        request:SignupRequest,
        session:HttpSession
    ):ResponseEntity<ApiResponse<Unit>>{
        val token = session.id
        session.setAttribute(token, request)
        memberService.verify(token, request)
        return ResponseEntity.ok(ApiResponse.noContent());
    }

    @GetMapping("exists/{id}")
    fun checkDuplicated(
        @PathVariable
        id:String):ResponseEntity<ApiResponse<Boolean>>{
        return ResponseEntity.ok(ApiResponse.success(memberService.checkId(id)))
    }

    @PreAuthorize("hasRole('ADMIN')")
    @GetMapping("{id}")
    fun get(
        @PathVariable
        id:String
    ):ResponseEntity<ApiResponse<MemberDetailResponse>>{
        val res = memberService.findById(id)
        return ResponseEntity.ok(ApiResponse.success(res))
    }
}